Scannable axes compute nullability and separability

rtjoa opened the pull request:
Work primarily by @zackbach.

This includes the following PRs, and is a follow-up to https://github.com/oxcaml/oxcaml/pull/5484.
1. https://github.com/oxcaml/oxcaml/pull/5031
2. https://github.com/oxcaml/oxcaml/pull/5086
3. https://github.com/oxcaml/oxcaml/pull/5095
4. https://github.com/oxcaml/oxcaml/pull/5128
5. https://github.com/oxcaml/oxcaml/pull/5170
6. https://github.com/oxcaml/oxcaml/pull/5178
7. https://github.com/oxcaml/oxcaml/pull/5187

ocamlformat PR: https://github.com/oxcaml/ocamlformat/pull/135

goldfirere commented on testsuite/tests/typing-jkind-bounds/poly-variant-limit/test_ikinds.ml:
```diff
@@ -29,9 +29,9 @@ type foo : immutable_data = Types.poly_variant_with_101
-       But the kind of type "Types.poly_variant_with_101" must be a subkind of
-           immutable_data
+       But the layout of type "Types.poly_variant_with_101" must be a sublayout of
+         value non_float
```
This seems like an error message regression

rtjoa replied:
We had error messages similarly change for `immediate`, which we addressed by adding the note below:
```
The layout of the first is value
  because of the definition of t_value at line 1, characters 0-12.
But the layout of the first must be a sublayout of value non_pointer
  because of the definition of t at line 2, characters 2-22.
Note: The layout of immediate is value non_pointer.
```
So analogously, added the following message in c2c64f0c035b6ec7533980c4829726ac6b51791b
```
Note: The kinds mutable_data, immutable data, and sync_data have
the layout value non_float.
```

goldfirere replied:
Just made a comment about this: https://github.com/oxcaml/oxcaml/commit/c2c64f0c035b6ec7533980c4829726ac6b51791b#r182043972

rtjoa replied:
Thanks, put a response in https://github.com/oxcaml/oxcaml/commit/c2c64f0c035b6ec7533980c4829726ac6b51791b#r182049206

goldfirere commented on testsuite/tests/typing-jkind-bounds/poly-variant-limit/test.ml:
```diff
@@ -28,9 +28,9 @@ type foo : immutable_data = Types.poly_variant_with_101
-       But the kind of type "Types.poly_variant_with_101" must be a subkind of
-           immutable_data
+       But the layout of type "Types.poly_variant_with_101" must be a sublayout of
+         value non_float
```
This seems like an error message regression

rtjoa replied:
(Responding to similar comment above)

goldfirere commented on testsuite/tests/typing-jkind-bounds/subsumption/basics.ml:
```diff
@@ -357,9 +357,9 @@ Error: Signature mismatch:
+       The layout of the first is value
          because of the definition of t at line 2, characters 2-36.
-       But the kind of the first must be a subkind of immutable_data with 'a
+       But the layout of the first must be a sublayout of value non_float
```
This error message seems worse than the old one

goldfirere commented on testsuite/tests/typing-jkind-bounds/subsumption/recursive_ikinds.ml:
```diff
@@ -20,30 +20,15 @@ type 'a my_list = Nil | Cons of 'a * 'a foo
-         No .cmi file found containing my_list.
-       A good next step is to add a layout annotation on a parameter to
-       the declaration where this error is reported.
+type 'a my_list =
```
This is newly accepted. Is that expected?

rtjoa replied:
This doesn't seem too surprising, given that
1. In the diff of the non-ikinds version of the test, scannable axes make the error change from complaining about the layout to complaining about the with bounds
2. So seems plausible that ikinds would take us the rest of the way there (I say, knowing little about ikinds).

(Also noting for redundancy, as the deleted CR says, it should be accepted!)

goldfirere commented on testsuite/tests/typing-jkind-bounds/verbosity/verbosity2.ml:
```diff
@@ -15,16 +15,14 @@ type t
           immutable
           portable
           contended
-          non_null
```
We're losing the `non_null` here. Should we be?

rtjoa replied:
Added printing of redundant nullability/separability on verbosity 2 in commit: 1dbc0629ad90ba74bda22b0c93851191458538a2

goldfirere commented on testsuite/tests/typing-jkind-bounds/verbosity/verbosity2.ml:
```diff
@@ -84,17 +78,13 @@ type t
   : value
       mod portable
-          non_null
-          separable
```
We're losing this `separable` too. (Normally, I'd say that's fine, but I see this file is named verbosity2.)

goldfirere commented on testsuite/tests/typing-jkind-bounds/basics.ml:
```diff
@@ -1649,25 +1651,6 @@ Error: This expression has type "int t" but an expression was expected of type
          because of the definition of require_portable at line 2, characters 21-57.
 |}]
 
-(*********************************)
```
Why delete this?

rtjoa replied:
Drive-by cleanup: it's a duplicate of test 17 below

goldfirere commented on testsuite/tests/typing-layouts/hash_types.ml:
```diff
@@ -478,32 +478,35 @@ and r = { x : int; y : float#; }
+       Note: The layout of immediate is value non_pointer.
 |}]
 
+(* CR layouts-scannable: improve this error message *)
```
Internal ticket to fix this?

rtjoa replied:
Internal ticket 6111 - added a reference

goldfirere commented on jane/doc/extensions/_03-unboxed-types/01-intro.md:
```diff
@@ -47,9 +38,67 @@ The type system also supports one *composite* layout: unboxed products:
+The `value` layout describes the representation of "vanilla" OCaml values. In OxCaml, we introduce finer distinctions within `value`, such as that some values are `non_pointer` and can thus get more efficient code generation; we also *broaden* the set of representations that the GC can handle, such as with `value maybe_null`, which contains all values in legacy OCaml plus the `NULL` pointer.
+
+The most general of the `value`-like layouts is `scannable`, whose only requirement of its elements is that they can be scanned by the GC.
+We call the sublayouts of `scannable` the "value layouts" (choosing to emphasize the more familiar layout, `value`, instead of the top of the lattice, `scannable`.)
```
```suggestion
We call the sublayouts of `scannable` the "value layouts" (choosing to emphasize the more familiar layout, `value`, instead of the top of the lattice, `scannable`).
```

goldfirere commented on jane/doc/extensions/_03-unboxed-types/01-intro.md:
```diff
@@ -47,9 +38,67 @@ The type system also supports one *composite* layout: unboxed products:
+The nullability axis records whether `NULL` (the machine word 0) is a possible
+value of a type, and is used to support the non-allocating option `'a or_null`
+type. The axis has two possible values, with `non_null < maybe_null`. A type may
+be `non_null` if none of its values are `NULL`.
```
```suggestion
be `non_null` only if none of its values are `NULL`.
```

goldfirere commented on jane/doc/extensions/_03-unboxed-types/01-intro.md:
```diff
@@ -47,9 +38,67 @@ The type system also supports one *composite* layout: unboxed products:
+- A type is `non_pointer` if none of its values are pointers (i.e., the bottom bit is tagged or the value is NULL).
+- A type is `non_pointer64` if none of its values are pointers, when compiled to native code for 64-bit systems.
+- A type is `non_float` if none of its values are floats.
+- A type is `separable` if either all or none of its values are floats. Separability is used to track types for which it is safe
```
```suggestion
- A type is `separable` if either all or none of its values are floats. In order for the float array optimization to be sound, only `separable` types may be stored in `array`s.
```

goldfirere commented on jane/doc/extensions/_03-unboxed-types/01-intro.md:
```diff
@@ -47,9 +38,67 @@ The type system also supports one *composite* layout: unboxed products:
+
+Scannable axes written after a value layout overwrite the previous value of the axis. They can also be written after non-value layouts, but have no effect, e.g. `float64 = float64 non_null = float64 maybe_null`. Scannable axes can also be written on `any`, in which case they take effect *only in the case* that it is lowered to a value layout. For example, `float64` and `value` are both sublayouts ...
+
+We also support the `mod` syntax for scannable axes, which only lowers the scannable axis (takes the meet). For example, `value mod non_pointer` is equivalent to `value non_pointer`, but `value mod maybe_separable` is equivalent to `value`.
```
```suggestion
We also support the `mod` syntax for scannable axes, which only lowers the scannable axis (takes the meet). For example, `value mod non_pointer` is equivalent to `value non_pointer`, but `value mod maybe_separable` is equivalent to `value`. (This support is meant to be transitional; we expect to remove this eventually. New code should write the scannable axes before the `mod`.)
```

goldfirere commented:
With rebasing, it can be hard to tell what to review and what not to review. You added a bunch of commits in response to my comments (thanks!). What should I review?

rtjoa commented:
> With rebasing, it can be hard to tell what to review and what not to review. You added a bunch of commits in response to my comments (thanks!). What should I review?

Ah, I should’ve specified - the commits in response to your review are the commits starting from “Add reference to internal ticket”. But the ones I think are actually worth reviewing are just those I’ve linked directly in my replies.

julesjacobs commented:
The ikind changes look good to me :+1:

julesjacobs commented:
Is the intent to make `mod` modifiers order independent?
The following example may still be order dependent:

```
type ('a : value mod non_pointer) require_non_pointer

type bad  : value mod non_pointer maybe_separable
type good : value mod maybe_separable non_pointer

type good_check = good require_non_pointer
type bad_check = bad require_non_pointer
```

zackbach commented:
Ideally, this example would give a warning—and does with the new warning 185 `overridden-kind-modifier` active (see `typing-layouts-scannable/warnings.ml`). This warning currently gets raised three times, due to three calls stemming from `transl_type_decl`, so it is disabled by default. See internal ticket 2801.

In the meantime, the last specified `mod` modifier overrides any previous ones (as in the example above). This seemed like reasonable short-term behavior while we're waiting on better error reporting… at least at the time.

EDIT: some of the information above has gone out-of-date, see Ryan's comments below (oops!)

julesjacobs commented:
Oh, I see, that makes sense, thanks!

julesjacobs commented:
What about these examples?

```
type a1 : value mod non_pointer maybe_separable
type a2 : value mod maybe_separable non_pointer

type c1 : value_or_null mod non_null maybe_null
type c2 : value_or_null mod maybe_null non_null

type e1 : value_or_null mod non_pointer maybe_separable
type e2 : value_or_null mod maybe_separable non_pointer
```

They output:

```
type a1
type a2 : value non_pointer
type c1 : value_or_null
type c2 : value maybe_separable
type e1 : value_or_null
type e2 : value_or_null non_pointer
```

but I am not seeing any warnings even with `-w +183..185`. I could easily be doing this wrong or misunderstanding, am I?
What is the short-term intended behavior and the eventual intended behavior?

rtjoa commented:
@julesjacobs Since the previous version of this PR, we changed the `mod` syntax back to meeting instead of overriding, for the motivation described in https://github.com/oxcaml/oxcaml/pull/5187.

However, it looks like I wrote a bug in that PR, which made it so multiple scannable axes under a single `mod` still override each other. Nice catch! I added a fix and test in 6f10084e2986d191c69a1b462d41d1dcc0ea3ede.

Regarding warnings - we just don't have warnings for scannable axes in mod bounds. The rationale was to simplify the code and avoid the triple-printing that @zackbach mentions, since we're deprecating the syntax and the only remaining need for mod-bounds is "weird" uses of ppx_template, which typically doesn't want warnings.

---

As an example of the intended behavior:
```
value a b mod c d mod e
```
get scannable axes
```
value_axes
|> override_with a
|> override_with b
|> meet_with [c; d; e]
```

zackbach commented:
Oh, this makes a lot of sense. My apologies for the stale information, and thanks Jules for the catch!

goldfirere commented on typing/jkind.ml:
```diff
@@ -3243,19 +3256,38 @@ module Violation = struct
         let should_check_jk2 = not print_as_value_layout in
-        ( check_has_component immediate_layout jk1
-          || (should_check_jk2 && check_has_component immediate_layout jk2),
+        let should_note_immediate =
```
Maybe write a sentence here about what you're trying to achieve at a high level? Otherwise we end up in a boolean soup (also true before these most recent changes).

rtjoa replied:
Good point, added some comments about error printing in dbe5cbf8115a13d68a7d7d6f84c04c87751d7693

goldfirere commented on typing/typemode.ml:
```diff
@@ -286,26 +292,26 @@ let transl_mod_bounds annots =
       | "non_pointer" ->
-        Transled_modifiers.set_separability bounds_so_far
-          (Some { txt = Non_pointer; loc })
+        Transled_modifiers.meet_separability bounds_so_far
```
These changes surprise me... I thought we wanted "set", not "meet".

rtjoa replied:
We do want "set" for `value x y z`. But this function translates mod bounds, and this code specifies how to fold over `x y z` in `value mod x y z` - as of https://github.com/oxcaml/oxcaml/pull/5187, we made scannable axes in mod bounds meet instead of override.
